dali/operators/decoder/audio/generic_decoder.cc

```diff
@@ -0,0 +1,230 @@
+          length_(length), curr_(0), input_(static_cast<const char *>(input)) {}
+
+
+  size_t length() {
+    return length_;
+  }
+
+
+  size_t seek(sf_count_t offset, int whence) {
+    switch (whence) {
+      case SEEK_SET:
+        curr_ = offset;
+        break;
+      case SEEK_CUR:
+        curr_ += offset;
+        break;
+      case SEEK_END:
+        curr_ = length_ + offset;
+        break;
+      default:
+        DALI_FAIL("Incorrect `whence` argument")
+    }
+    return curr_;
+  }
+
+
+  size_t read(void *dst, sf_count_t num) {
+    num = std::min<sf_count_t>(num, length_ - curr_);
+    memcpy(dst, input_ + curr_, num);
+    curr_ += num;
+    return num;
+  }
+
+
+  size_t tell() {
+    return curr_;
+  }
+
+
+ private:
+  size_t length_, curr_;
+  const char *input_;
+};
+
+/*
+ * Functions: `GetFileLen`, `Seek`, `Read`, `Tell` are the callbacks required
+ * by `sf_open_fd` interface. Descriptions of these functions are copy-pasted
+ * from `libsnd` docs.
+ */
+
+/**
+ * The virtual file context must return the length of the virtual file in bytes.
+ */
+sf_count_t GetFileLen(void *self) {
+  return static_cast<MemoryStream *>(self)->length();
+}
+
+
+/**
+ *  The virtual file context must seek to offset using the seek mode provided by whence
+ *  which is one of
+ *   SEEK_CUR
+ *   SEEK_SET
+ *   SEEK_END
+ * The return value must contain the new offset in the file.
+ */
+sf_count_t Seek(sf_count_t offset, int whence, void *self) {
+  return static_cast<MemoryStream *>(self)->seek(offset, whence);
+}
+
+
+/**
+ * The virtual file context must copy ("read") "num" bytes into the buffer provided by dst
+ * and return the count of actually copied bytes.
+ */
+sf_count_t Read(void *dst, sf_count_t num, void *self) {
+  return static_cast<MemoryStream *>(self)->read(dst, num);
+}
+
+
+/**
+ * Return the current position of the virtual file context.
+ */
+sf_count_t Tell(void *self) {
+  return static_cast<MemoryStream *>(self)->tell();
+}
+
+}  // namespace
+
+
+template<typename SampleType>
+void GenericAudioDecoder<SampleType>::DecodeTyped(span<SampleType> output) {
+  impl_->DecodeTyped(output);
+}
+
+
+template<typename SampleType>
+AudioMetadata GenericAudioDecoder<SampleType>::OpenImpl(span<const char> encoded) {
+  return impl_->OpenImpl(encoded);
+}
+
+
+template<typename SampleType>
+void GenericAudioDecoder<SampleType>::CloseImpl() {
+  impl_->CloseImpl();
+}
+
+
+template<typename SampleType>
+GenericAudioDecoder<SampleType>::~GenericAudioDecoder() = default;
+
+
+template<typename SampleType>
+GenericAudioDecoder<SampleType>::GenericAudioDecoder() :
+        impl_(std::make_unique<Impl>()) {
+}
+
+
+template<typename SampleType>
+struct GenericAudioDecoder<SampleType>::Impl {
+  void DecodeTyped(span<SampleType> output) {
+    read_samples(snd_file_, output);
+  }
+
+
+  AudioMetadata OpenImpl(span<const char> encoded) {
+    assert(!encoded.empty());
+    assert(encoded.data());
```

Comment:
Is it possible that `!encoded.empty()` but not `encoded.data()`?

Reply:
Changed to `assert(!encoded.empty());`